### DISCUSSION TITLE
NLP_fine-tuning for Question Answering

### DISCUSSION THREAD

**Discussion by seuly1203** [O-per/cakd3_Project3#9] [Study & Ideas]
[BERT를 사용한 QA 설명 및 예제코드(영문)](https://towardsdatascience.com/question-answering-with-a-fine-tuned-bert-bc4dafd45626)
[KorQuAD 2.0 한국어 질의응답 데이터셋](https://korquad.github.io/)
[Building a QA System with BERT on Wikipedia (영문) - 과정에 대한 상세한 설명](https://qa.fastforwardlabs.com/pytorch/hugging%20face/wikipedia/bert/transformers/2020/05/19/Getting_Started_with_QA.html)
[BERT를 사용한 질의응답 시스템 프로젝트(국문)](https://githubmemory.com/repo/JoungheeKim/korean-question-answer-system)

**Comment by seuly1203** [comment]
[BERT]

BERT에 대해 자세히 설명된 PPT[(by 김현기/한국전자통신연구원)](http://software.kaist.ac.kr/wp-content/uploads/2020/06/200618-.pdf)
KorBERT - QA 미세조정 관련 PPT[(by 임준호/한국전자통신연구원)](https://www.slideshare.net/LGCNSairesearch/nlu-tech-talk-with-korbert)
[KoBERT-Transformers ](https://github.com/monologg/KoBERT-Transformers )

**Reply by seuly1203** [reply]
기계독해 - 형태소기반 korbert 사용했을 때 EM 및 F1 score 가장 높게 나타남

**Reply by seuly1203** [reply]
KorBERT morph 모델(1 tensorflow,2 pytorch) 사용하기

**Comment by seuly1203** [comment]
[HuggingFace 🤗]
 [허깅페이스 퀵스타트 (영문)](https://huggingface.co/transformers/v2.4.0/quickstart.html)
[트랜스포머 노트북(영문)](https://huggingface.co/transformers/notebooks.html)

**Comment by yibnn** [comment]
참고링크
https://choice-life.tistory.com/45

**Comment by Kyounglim-Joey** [comment]
Bert For Question Answering (logic)
https://medium.com/analytics-vidhya/bert-for-question-answering-cd00741d1dab
* 다음 링크를 통해 기존에 없는 KoBERT의 BertFORQuestionAnswering을 사용할 수 있지않을까 해서 공유합니다.

**Comment by hye-jj** [comment]
BERT 한국어버전(korquad) training 및 evaluating 해보기
https://buttercoconut.xyz/323/

**Comment by hye-jj** [comment]
KoBERT 쉽게 따라하고 간단한 fine-tuning 하기
https://moondol-ai.tistory.com/241

**Comment by hye-jj** [comment]
트랜스포머모델을 도입했고 여기에 한국어 형태소 분석기 **MeCab**
을 적용해 단어처리의 어려움을 완화했다.
https://www.koreascience.or.kr/article/JAKO201931765017930.pdf

[Technology Toolkit 2021]
스마트한 텍스트 분석을 향한 첫걸음, KoreALBERT!!

한국어 사전학습 모델
영어 기반의 다양한 모델들이 공개되었지만, 한국어의 경우 영어와는 다른 언어적 특성이 있기 때문에 영어 모델을 그대로 활용하기는 어렵습니다. 삼성SDS 연구소는 많은 양의 한글 데이터를 직접 수집하여 사전학습 모델을 만들었습니다.

① KoreALBERT

첫 번째로 소개할 모델은 KoreALBERT(Korea + ALBERT)입니다[그림 6]. BERT의 경량화 모델인 ALBERT(A Lite BERT) 아키텍처를 기반으로 한글을 학습시켰습니다. 학습 데이터는 다양한 분야의 비즈니스에 공통으로 적용이 가능하도록 한글 위키피디아, 신문 기사, 책 줄거리 등 일반적인 분야의 문어체 위주 한글 텍스트로 구성되었습니다. 또한 기존에 존재하는 다른 한글 언어 모델과 차별화하기 위해 모델 경량화에 강점을 갖는 새로운 아키텍처와 특허를 받은 학습 방법을 사용하였습니다(3. 차별화 포인트).

② KoELECTRA

ELECTRA는 BERT 이후에 등장한 언어모델로서, BERT가 가진 학습 데이터 사용의 비효율성을 극복하기 위해 탄생한 모델입니다. 기존 모델인 BERT는 학습 과정에서 전체 입력 토큰 중 [MASK]로 가려진 15%의 토큰들만 학습에 사용하기 때문에 데이터 효율성이 떨어지게 됩니다. ELECTRA는 이를 극복하기 위해 [MASK]로 가려지지 않은 나머지 85% 토큰에 대해서도 학습을 진행하므로 BERT 대비 초기 학습 속도와 성능 면에서 우수하다고 증명되었습니다. 이러한 ELECTRA 모델 아키텍처를 한글로 학습한 모델이 KoELECTRA입니다. KoreALBERT보다 무겁지만, 더 좋은 성능을 자랑합니다.

#### Use Cases
Case 1 : 기계독해(MRC, Machine Reading Comprehension)
지문을 이해하고 함께 입력된 질문에 대한 답을 찾는 기능을 수행할 수 있습니다. 기계독해 태스크를 수행하기 위해서 [그림 11]과 같이 사전학습 모델을 베이스라인 모델로 사용하였고, Enhanced CNN을 추가하여 기계독해 모델의 성능을 향상했습니다. 이러한 모델을 사용하여 KorQuAD 1.0 에서 좋은 성적을 거두었습니다.

Case 2 : 텍스트 분류(Text Classification)
텍스트 분류는 입력된 텍스트가 어떤 카테고리에 해당하는지 분류하는 태스크입니다. 예를 들어, 입력 텍스트가 긍정적인지 부정적인지를 판별하는 감정 분류(Sentiment Classification)를 할 수 있습니다. 감정 분류는 사용자 댓글이나 제품 피드백을 분류하는 데 많이 사용합니다. 또한 [그림 12]와 같이 텍스트에 어떤 의도가 있는지 판별(Intent Classification)하는 데도 사용할 수 있습니다. 그뿐만 아니라, 어떤 텍스트든 정해놓은 특정 카테고리로 분류할 수 있어서 실제로 가장 많이 활용하는 태스크입니다. 텍스트 분류 모델 또한 사전학습 모델(KoreALBERT)에 따라 전체 성능이 달라집니다.

Case 3 : 문장 유사도 분석(STS, Semantic Textual Similarity)
문장 유사도 분석은 두 문장이 의미적 혹은 문맥적으로 얼마나 유사한지를 분석하는 태스크입니다. 이 태스크는 질의 문장과 저장된 문장 중 가장 비슷한 문장을 찾아내는 문장 검색 기능으로 활용할 수 있습니다. 문맥적인 부분까지 고려하기 때문에 기존 키워드 검색보다 감성 품질이 더 우수하다는 강점이 있습니다[그림 13 상].

검색해야 하는 문장의 개수가 늘어날수록 처리 시간이 급격히 길어집니다. 삼성SDS 연구소는 이러한 문제를 해결하기 위해 샴 네트워크(Siamese Neural Network)와 CNN 네트워크를 활용하는 새로운 아키텍처를 제안했습니다[그림 13 하]. 해당 연구 결과는 특허 등록과 해외 인공지능학회인 ICPR 2020에 발표하였습니다.

Case 4 : 문서 요약 모델(Text Summarization)
문서 요약 모델은 [그림 14]와 같이 문서의 주요 문장을 추출하여 요약 문장을 만드는 추출 요약 모델(Extractive Summarization)과 문서의 내용을 이해하고 문맥에 맞는 문장을 생성하는 생성 요약 모델(Abstractive Summarization)이 있습니다. 각각 추출 요약은 NLU, 생성 요약은 NLG를 기반으로 모델을 만들 수 있습니다. 문서 요약 기능은 많은 문서를 통해 인사이트를 얻어야 하는 분석가들에게 요약된 내용을 전달함으로써 시간과 비용을 줄여 줄 수 있습니다.

영상 : https://youtu.be/5SXlNvM3jKM
원문 : https://www.samsungsds.com/kr/insights/TechToolkit_2021_KoreALBERT.html

**Comment by sumunoh** [comment]
기계독해 관련 논문들
[정보 검색 기반 기계독해 시스템을 위한 단락 재순위화 모델.pdf](https://github.com/O-per/cakd3_Project3/files/7576093/default.pdf)
[KorQuAD 2.0 웹문서 기계독해를 위한 한국어 질의응답 데이터셋.pdf](https://github.com/O-per/cakd3_Project3/files/7576094/KorQuAD.2.0.pdf)
[BERT를 이용한 한국어 질의응답 데이터 셋에서의 기계 독해.pdf](https://github.com/O-per/cakd3_Project3/files/7576095/BERT.pdf)
[다중 작업 학습을 통한 문장 유사도 기반 단락 재순위화 방법.pdf](https://github.com/O-per/cakd3_Project3/files/7576096/default.pdf)

**Comment by seuly1203** [comment]
출처: https://github.com/JoungheeKim/korean-question-answer-system
[한국어 Question & Answering 시스템 개발 - 중간발표](https://www.youtube.com/watch?v=fGQAx_wCm3E&ab_channel=%EA%B9%80%EC%A0%95%ED%9D%AC)
